### PR TITLE
1080 merge harvard

### PR DESCRIPTION
**Pull request by flooie**
New Branch, builds off tax opinions to merge Harvard data into cl.

1) Database model changes include numerous additional fields as found in the meta data of the caselaw project.
   New fields in Docket, Opinion, OpinionCluster
2) A new command/management in corpus importer
3) Minor change to view_opinions template to allow Harvard html to be viewed on cl.

### CODE REVIEW

File: cl/search/models.py [hunk 1]
```diff
@@ -1896,6 +1902,72 @@ class OpinionCluster(models.Model):
...
+        "headnotes. These are added later by an editor not "
+        "connected to the court, but who instead works for a "
+        "legal publishing house. ",
+        blank=True,
+    )
+    summary = models.TextField(
+        help_text="A summary of what happened in the case. "
+        "Appears at the beginning of the case just "
+        "after the title of the case and court "
+        "information. Not identified as such.",
```
Comment (mlissner) [inline cl/search/models.py]: I'm not sure what "not identified as such" means in this context.

File: cl/search/models.py [hunk 2]
```diff
@@ -1896,6 +1902,72 @@ class OpinionCluster(models.Model):
...
+        blank=True,
+    )
+    history = models.TextField(
+        help_text="History of the case (similar to the summary, "
+        "but focused on past events related to this case). "
+        "Appears at the beginning of the case just after "
+        "the title of the case and court information",
+        blank=True,
+    )
+    other_date = models.TextField(
```
Comment (mlissner) [inline cl/search/models.py]: Should this be `other_dates`? I thought it could contain more than one?

File: cl/search/models.py [hunk 3]
```diff
@@ -1896,6 +1902,72 @@ class OpinionCluster(models.Model):
...
+    correction = models.TextField(
+        help_text="Publisher's correction to the case text. "
+        "Example: Replace last paragraph on page 476 "
+        "with this text: blah blah blah. This is basically an"
+        " unstructured text that can be used to manually "
+        "correct case content according to publisher's "
+        "instructions. No footnotes is expected within it.",
+        blank=True,
+    )
+    html_harvard = models.TextField(
```
Comment (mlissner) [inline cl/search/models.py]: Sorry, I should be better with their API, but is this the html as in the field name or the XML as in the description? Can you clarify somehow?

File: cl/search/models.py [hunk 4]
```diff
@@ -1923,6 +1995,27 @@ class OpinionCluster(models.Model):
         db_index=True,
         default=False,
     )
+    filepath_local = models.FileField(
+        help_text="Path to Harvard json local storage.",
```
Comment (mlissner) [inline cl/search/models.py]: Let's make this description generalizable. Something more like, "The source file for the entire cluster. This is used by the Harvard data. See also Opinion.filepath_local for..."

File: cl/search/models.py [hunk 5]
```diff
@@ -1923,6 +1995,27 @@ class OpinionCluster(models.Model):
...
+        "binary file or text data",
+        max_length=40,
+        db_index=True,
+    )
+    page_count = models.IntegerField(
+        help_text="The number of pages in the opinion cluster.",
+        blank=True,
+        null=True,
+    )
+    image_missing = models.BooleanField(
```
Comment (mlissner) [inline cl/search/models.py]: Are we using this or do we plan to? Bit of a curious one.

File: cl/search/models.py (hunk 3 above)
Comment (mlissner) [reply cl/search/models.py]: Hm, I see now that we have this field on the opinion object as well. That's...curious and a tad confusing. Maybe a better name for one or the other would make sense.

File: cl/search/models.py [hunk 6]
```diff
@@ -58,6 +58,8 @@
     ("ZR", "columbia merged with resource.org"),
     ("ZCR", "columbia merged with court and resource.org"),
     ("ZL", "columbia merged with lawbox"),
+    ("U", "Harvard, Library Innovation Lab Case Law Access Project"),
+    ("CU", "court website merged with Harvard"),
```
Comment (mlissner) [inline cl/search/models.py]: The dumb thing about this scheme is that we're going to need all the other "Combined with" possibilities. This should eventually be a m2m field, but until we can do that, we're stuck with this dumb (old) system that I invented. We can do all of those other possibilities now or later....

File: cl/opinion_page/templates/view_opinion.html [hunk 7]
```diff
@@ -394,6 +394,8 @@ <h4>Summary of Opinion</h4>
                                     {{ sub_opinion.html_columbia|safe }}
                                 {% elif sub_opinion.html_lawbox %}
                                     {{ sub_opinion.html_lawbox|safe }}
+                                {% elif sub_opinion.html_harvard %}
```
Comment (mlissner) [inline cl/opinion_page/templates/view_opinion.html]: I assume you've taken a look at this and it looks good, great, or at least passable?

File: cl/corpus_importer/management/commands/harvard_opinions.py [hunk 8]
```diff
@@ -0,0 +1,351 @@
...
+
+        # Handle duplicate citations.  By comparing date filed and page count
+        # I find it unlikely two cases would both start and also stop on the
+        # same page.  So we use page count as a proxy for it.
+        if cite_search.count() > 0:
+            cluster_id = cite_search[0].cluster_id
+            cluster = OpinionCluster.objects.filter(id=cluster_id)[0]
+            if cluster.date_filed is not None:
+                date_filed, is_approximate = validate_dt(data["decision_date"])
+                if str(cluster.date_filed) != str(date_filed):
```
[the remark this answers is not in the document]
Comment (flooie) [reply cl/corpus_importer/management/commands/harvard_opinions.py]: √

File: cl/corpus_importer/management/commands/harvard_opinions.py [hunk 9]
```diff
@@ -0,0 +1,351 @@
...
+            Citation.objects.all()
+            .filter(reporter=rep)
+            .filter(page=page)
+            .filter(volume=vol)
+        )
+
+        pg_count = 1 + int(data["last_page"]) - int(data["first_page"])
+
+        # Handle duplicate citations.  By comparing date filed and page count
+        # I find it unlikely two cases would both start and also stop on the
```
[the remark this answers is not in the document]
Comment (flooie) [reply cl/corpus_importer/management/commands/harvard_opinions.py]: √

File: cl/corpus_importer/management/commands/harvard_opinions.py [hunk 10]
```diff
@@ -0,0 +1,351 @@
...
+        # same page.  So we use page count as a proxy for it.
+        if cite_search.count() > 0:
+            cluster_id = cite_search[0].cluster_id
+            cluster = OpinionCluster.objects.filter(id=cluster_id)[0]
+            if cluster.date_filed is not None:
+                date_filed, is_approximate = validate_dt(data["decision_date"])
+                if str(cluster.date_filed) != str(date_filed):
+                    logger.info("Duplicate cite string different date filed")
+                elif cluster.page_count != pg_count:
```
[the remark this answers is not in the document]
Comment (flooie) [reply cl/corpus_importer/management/commands/harvard_opinions.py]: Actually no.  We go here if we have a duplicate cite.  that duplicate cite is presumably just been added by the merge and we are getting the next in the list which has a page count.

File: cl/corpus_importer/management/commands/harvard_opinions.py [hunk 11]
```diff
@@ -0,0 +1,351 @@
...
+        # Flag them for a later crawl by using the placeholder '[[Image]]'
+        missing_images = True if content.find("[[Image here]]") > -1 else False
+        judge_list = [
+            find_judge_names(x.text) for x in soup.find_all("judges")
+        ]
+        author_list = [
+            find_judge_names(x.text) for x in soup.find_all("author")
+        ]
+        judges = ", ".join(
+            list(set(itertools.chain.from_iterable(judge_list + author_list)))
```
[the remark this answers is not in the document]
Comment (flooie) [reply cl/corpus_importer/management/commands/harvard_opinions.py]: This creates the entire list of judges on the case, the author is often not in the, other judges on this case list.

File: cl/corpus_importer/management/commands/harvard_opinions.py [hunk 12]
```diff
@@ -0,0 +1,351 @@
...
+        while json_fields:
+            cd = {}
+            key = json_fields.pop(0)
+            iterations = 0
+            for item in [x.text for x in soup.find_all(key)]:
+                cd[str(iterations)] = item
+                iterations += 1
+            if cd == {}:
+                cd = ""
+            data_set[key] = cd
```
[the remark this answers is not in the document]
Comment (flooie) [reply cl/corpus_importer/management/commands/harvard_opinions.py]: This was meant to be a placeholder for this data set - until we had an opportunity to go back and parse through the data.

File: cl/corpus_importer/management/commands/harvard_opinions.py [hunk 13]
```diff
@@ -0,0 +1,351 @@
...
+        docket_dictionary = {
+            "case_name": data["name"],
+            "docket_number": docket_string,
+            "court_id": court,
+            "source": Docket.HARVARD,
+            "ia_needs_upload": False,
+            "slug": slugify(data["name"]),
+        }
+        logger.info("Adding docket for: %s", cite)
+        docket = Docket.objects.create(**docket_dictionary)
```
[the remark this answers is not in the document]
Comment (flooie) [reply cl/corpus_importer/management/commands/harvard_opinions.py]: correct.

File: cl/corpus_importer/management/commands/harvard_opinions.py [hunk 14]
```diff
@@ -0,0 +1,351 @@
...
+            "correction": data_set["correction"],
+            "judges": judges,
+            "html_harvard": str(soup),
+            "sha1": sha1_of_json_data(json.dumps(data)),
+            "page_count": pg_count,
+            "image_missing": missing_images,
+            "filepath_local": file_path,
+        }
+        logger.info("Adding cluster for: %s", cite)
+        cluster = OpinionCluster.objects.create(**cluster)
```
[the remark this answers is not in the document]
Comment (flooie) [reply cl/corpus_importer/management/commands/harvard_opinions.py]: √

File: cl/corpus_importer/management/commands/harvard_opinions.py [hunk 15]
```diff
@@ -0,0 +1,351 @@
...
+            opinion_data = {
+                "type": op_type,
+                "cluster_id": cluster.id,
+                "author_str": author_str,
+                "download_url": ia_download_url,
+                "html_harvard": str(op),
+                "joined_by_str": joined_by_str,
+            }
+            logger.info("Adding opinion for: %s", cite)
+            Opinion.objects.create(**opinion_data)
```
[the remark this answers is not in the document]
Comment (flooie) [reply cl/corpus_importer/management/commands/harvard_opinions.py]: just seems easier to read but I appreciate the other approach.

File: cl/search/models.py (hunk 5 above)
Comment (flooie) [reply cl/search/models.py]: Yes, we need it to know which cases to re-scrape at a later date.

File: cl/search/models.py (hunk 6 above)
Comment (flooie) [reply cl/search/models.py]: Either way works for me.

File: cl/opinion_page/templates/view_opinion.html (hunk 7 above)
Comment (flooie) [reply cl/opinion_page/templates/view_opinion.html]: Yes, It views normally to my eyes and my computer.

File: cl/search/models.py (hunk 3 above)
Comment (flooie) [reply cl/search/models.py]: Yeah, I guess this should read xml_harvard.

Comment (flooie) [reply cl/search/models.py]: That should help when we actually get the html_harvard for some other cases.

File: cl/corpus_importer/management/commands/harvard_opinions.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (flooie) [reply cl/corpus_importer/management/commands/harvard_opinions.py]: I am flattening and deduping required for data returned from find_judge_names

File: cl/search/models.py (hunk 2 above)
Comment (flooie) [reply cl/search/models.py]: changed

File: cl/corpus_importer/management/commands/harvard_opinions.py [hunk 16]
```diff
@@ -319,6 +319,110 @@ def parse_harvard_opinions(reporter, volume):
             replace("Docket No.", ""). \
             replace("Docket Nos.", "")
         with transaction.atomic():
```
Comment (mlissner) [inline cl/corpus_importer/management/commands/harvard_opinions.py]: Next time we do something like this, we should see how hard it is to do:

```
d = Docket(var=foo, zoo=boo)
cl = Cluster(asdf=asdf, asdf=asdf)
op = Opinion(blah, blah)
with transaction.atomic():
    d.save()
    cl.save()
    op.save()
```

I don't want to bother with this now, but indenting 100 lines to accomplish this feels icky.

File: cl/corpus_importer/management/commands/harvard_opinions.py [hunk 17]
```diff
@@ -159,17 +159,14 @@ def parse_harvard_opinions(reporter, volume):
...
-                if cluster.date_filed != date_filed:
-                    logger.info("Duplicate cite string different date filed")
-                elif cluster.page_count != pg_count:
-                    logger.info("Duplicate cite string but diff page count")
-                else:
-                    logger.info("%s Already in CL." % cite)
-                    continue
+            cluster_ids = cite_search.values_list("cluster_id")
+            objects = OpinionCluster.objects.filter(
+                id__in=cluster_ids).values_list('case_name')
```
Comment (mlissner) [inline cl/corpus_importer/management/commands/harvard_opinions.py]: Calling a variable `objects` is a no-go. But beyond that, there's hardly ever a reason to look up items by ID like this. I rewrote these.

**Comment by mlissner** [comment]
Comments from our call:

```
1. Model changes (Bill):
    - Remove xml_harvard from cluster
    - Change cluster.filepath_local --> cluster.json_harvard
    - Remove sha1 from cluster
    - Remove page_count (for now)
    - Remove image_missing (for now, use Solr later if needed)
    - Fix description of opinion.xml_harvard
    - Make sure all Text/Char fields have blank=True, and do not have null=True

1. Stop putting IA info into the opinion.download_url field

1. Refactor harvard_opinions.py:205-211
   
1. File bug against Harvard that API results lack images
```

**Pull request merged**